### DISCUSSION TITLE
Custom processing option for mdx loader

### DISCUSSION THREAD

**Discussion by benjaminpreiss** [mdx-js/mdx#2134] [Ideas] [answered]
Following [a previous discussion on sections in mdx](https://github.com/mdx-js/mdx/issues/454#issuecomment-1242894555) I would like to propose an option in mdx loader that allows for custom logic in the processing of MDX files.

Let's assume the example of MDX with sections for different content blogs:

```js
{/* c:how-it-works.main */}

Some MDX here
...

{/* c:start */}

Some more MDX here
...
```

One might want to split process the above file and later receive an object as a default export, that would look something like this:

```js
const res = {
	"some.content.path": function() {
		const {Fragment: _Fragment, jsx: _jsx} = arguments[0]
		const no = 3.14
		function _createMdxContent(props) { /* … */ }
		function MDXContent(props = {}) { /* … */ }
		return {no, default: MDXContent}
	},
	"some.other.content.path": function() {
		//...
	},
	//...
};
export default res;
```

To achieve this, I propose we add an option to mdx loader that allows custom processing of files ([as implemented here](https://github.com/benjaminpreiss/mdx/blob/main/packages/loader/lib/index.js#L67-L81)):

```js
// ...
if(options.execute) {
    options.execute.call(this, value, process, callback)
  } else {
    process({value, path: this.resourcePath}).then(
      (file) => {
        callback(null, file.value, file.map)
      },
      (/** @type VFileMessage */ e) => {
        const fpath = path.relative(this.context, this.resourcePath);
        e.message = `${fpath}:${e.name}: ${e.message}`;
        callback(e);
      }
    )
  }
// ...
```

This gives the user the freedom to pass the following as config to the loader in webpack:

```js

function execute(value, process, callback) {
  const splittedFile = value.split(/{[\n\t ]*\/\*\s*c:[\w.\-\\[\d\]]+\s*\*\/[\n\t ]*}/)
  splittedFile.shift()
  const matchingComments = value.match(/(?<={[\n\t ]*\/\*\s*c:)[\w.\-\\[\d\]]+(?=\s*\*\/[\n\t ]*})/g)

  if(splittedFile.length) {
    Promise.all(splittedFile.map(async (v, i) => {
      const file = await process({value: v, path: this.resourcePath})
      return file
    })).then((arr) => {
        const objElements = arr.map((file, i) => {
          return `"${matchingComments[i]}": function() {
            ${file.value}
          },`
        })
        const result = `
          const res = {
            ${objElements.join("")}
          };
          export default res;
        `
        callback(null, result, arr[0].map)
        return result
    }, callback)
  } else {
    process({value, path: this.resourcePath}).then((file) => {
      const result = `
        export default function() {
          ${file.value}
        }
      `
      callback(null, result, file.map)
      return result
    }, callback)
  }
}

// ...
// webpack config excerpt:

      {
        test: /\.mdx$/,
        use: [
          {
            loader: path.resolve(__dirname, 'loaders/mdx-extended-loader/index.cjs'),
            options: {
              execute: execute,
              outputFormat: "function-body"
            }
          }
        ],
      }
// ...
```

**Comment by ChristianMurphy** [comment, marked as answer]
Heya @benjaminpreiss!
Exciting to see progress towards multiple documents in a file.
Using comments to provide metadata for the sections is interesting, could be a valuable addition.

I'd second @wooorm's comment

> Regexing files and injecting JS-like things as strings seems like a fragile way to go about it 

This feels like something which may be better served working off of a syntax tree.
Whether that be the initial markdown AST (MDAST/remark), the intermediate HTML AST (HAST/rehype), or the final JavaScript AST (ESAST/recma).
Which mdx already supports plugins for (https://github.com/mdx-js/mdx/tree/main/packages/mdx#optionsremarkplugins) which could make generating a proof of concept easier.
And would potentially remove the need for adding a new option to the loader.

**Reply by benjaminpreiss** [reply]
Ah, okey - sure! If that is possible with a remark plugin, could you give me a hint how? I'd be glad to implement a working alternative...

Just to be very clear about what I want to achive:

I want to be able to selectively import only certain sections of the MDX by their identifier (which is specified in specially formatted comments). These sections could then hold a variety of things, like exported variables or markdown. See this excerpt from the filecoin saturn page as an example of a MDX file:

```
{/* c:start */}

### Filecoin Saturn

Saturn is an open-source, decentralized Content Delivery Network (CDN) built on Filecoin.  

export const ctas = [
    {text: "Get Involved", link: "#getinvolved"},
    {text: "Learn more", link: "#whatissaturn"}
]

{/* c:what-is-saturn.title */}

# What is Saturn?

{/* c:what-is-saturn.description */}

Saturn is a Web3 CDN in Filecoin’s retrieval market. On one side of the network, websites buy fast, low-cost content delivery. On the other side, Saturn node operators earn Filecoin by fulfilling requests.

Saturn is trustless, permissionless, and inclusive. Anyone can run Saturn software, contribute to the network, and earn Filecoin.

Content on Saturn is [IPFS](https://ipfs.io/) content-addressed. Every piece of content is immutable and every response verifiable.

Crypto-incentives unite, align, and grow the network. Node operators earn Filecoin for accelerating web content and websites get faster content delivery for less.

All work is open source and developed in the public for all humanity.
```

In that case I would like to be able to use only the "ctas" array from the "start" section...

Would that be possible with a remark plugin?

**Reply by ChristianMurphy** [reply]
> sure! If that is possible with a remark plugin, could you give me a hint how? 

For getting started with ASTs I'd recommend reading through https://unifiedjs.com/learn/
The plugin would most likely want to work with JavaScript/ESAST rather than directly with Markdown/MDAST, since you're aim is to generate specific identifiers at the end.

The gist of an AST approach would be to:
1. find all the breaks or comments in an MDX document (whichever is being used to separate documents)
2. Identify the nodes between these separators, and keep track of them as sections
3. Generate new AST nodes to create the exports for each section, replacing the original ESAST document
   * Ideally https://github.com/tc39/proposal-js-module-blocks or https://github.com/tc39/proposal-module-fragments down the road, but probably not yet, as they aren't officially part of the ECMA standard.
   * Currently it could do this by creating blocks or IIFEs, and replacing `export` with `return`
   * Or it could keep the flattened structure (rather than nesting objects) and prepend a prefix to export names which make it clear which page/section they came from

Some reference tools:
* The playground gives an easy way to see the structure of ESAST https://mdxjs.com/playground/
* @remcohaszing created a remca plugin which gives an example of how to analyze JS scope with [periscopic](https://github.com/Rich-Harris/periscopic) and modify the final AST https://github.com/remcohaszing/recma-nextjs-static-props which can give some insights on how to structure a recma plugin.
* https://github.com/mrzmmr/unist-util-find-all-between offers a utility to find everything between two AST nodes.

**Comment by wooorm** [comment]
> For my purpose, sadly the `<Section>` tags are not sufficient, because I want to populate a global "content" object with different keys containing different sections of my page.

Reading this, it sounds like you interpreted my statements as saying that you should write `<Section>`s. That is not what I meant. I meant to say that I think it wise to build an AST from your comments as those sections.

Because most of the work is turning separate sections, whether with tags, thematic breaks, or with comments, into different components. Splitting these two steps allows both your custom syntax and other syntaxes.

The next steps are what @ChristianMurphy describes above. But I think it’s easier to perform those steps on an AST that‘s split in sections.

**Reply by benjaminpreiss** [reply]
Okey, thanks for the clarification! Sounds good to me 😄

**Reply by wooorm** [reply]
Sweet :) Let us know if you need help!

**Comment by karlhorky** [comment]
As mentioned in https://github.com/mdx-js/mdx/issues/454, this would also be super useful for building slide decks as [`mdx-deck`](https://github.com/jxnblk/mdx-deck), [`next-mdx-deck`](https://github.com/whoisryosuke/next-mdx-deck) and [`MDXP`](https://github.com/0phoff/MDXP) have done.

@benjaminpreiss maybe you could also add some content from #454 in your original description above as another use case for this?

Would be great if this were published as a package with configurable delimiters! (eg. in presentations, having the ability to separate pages with a simple `---` would be nice). If published, this would be a great library to build a simpler presentation library upon.

**Comment by benjaminpreiss** [comment]
Okey, I am thrilled to anounce that I am finally building it!

Here is the future repo for reference: https://github.com/frontline-hq/recma-sections

I have a question though - working with the estrees `.start`, `.end`, `.loc` and `.range` properties is really tiresome when inserting / manipulating stuff :/ . Here an example code snippet for comparison:

```js
"id": {
        "type": "Identifier",
        "start": 87,
        "end": 91,
        "name": "blue"
      }
```

What are these properties for? Only for source maps? Cause in that case I wouldn't change them...

**Reply by wooorm** [reply]
They are for positional info.
If you are *generating* nodes that weren’t in the original document, they **must not have** positional info.

**Reply by remcohaszing** [reply]
It’s used for various tools that work with the AST. Indeed, for source maps (I think), but also for the JSX automatic dev runtime, or potential MDX linters.

Typically you ignore it. It should only exist for nodes that are based on content in the original document, so generated nodes should not include this.

**Reply by benjaminpreiss** [reply]
Hmmm, okey! But what should I do with nodes that I replace with other nodes? A typical result would then be a path in the tree where in the middle of the path the positional information goes missing

**Reply by remcohaszing** [reply]
The new node is still a generated node, so it shouldn’t have positional information. Its children may have been relocated, but should keep their original positional information, so they could potentially be mapped back to the source code.

**Comment by benjaminpreiss** [comment]
Okey, first version is pushed... Please have a look and let me know what you think!

https://github.com/frontline-hq/recma-sections

**Reply by benjaminpreiss** [reply]
@karlhorky Is this sufficient for your usecase?

**Reply by karlhorky** [reply]
Thanks for this! From a quick glance at the code and readme, it seems that it splits an MDX file into multiple MDX components, so it seems indeed helpful for our use case to create a new version of `mdx-deck`!

We'll probably want to configure the delimiter to be `---` (`<hr />` element), but this should be possible.

cc @ProchaLu @EricCote this may provide some basis for [the `remark-mdxs` package that you're trying to build for MDX v2](https://github.com/mdx-js/mdx/pull/507#discussion_r1053332632)

**Comment by benjaminpreiss** [comment]
There is now an official npm release, typescript support and compilation to cjs and esm: https://www.npmjs.com/package/@frontline-hq/recma-sections

**Reply by wooorm** [reply]
Hi! Great work!

It’s cool that you got this working.
The output doesn’t look as clean as it could be, but if it works, it works.
Some quick ideas for improving that:

* No need to import the runtime again
* Perhaps you can hoist imports to the top?
* It’s a bit weird how all the components still have whitespace in them left over
* I recommend [`Rich-Harris/estree-walker`](https://github.com/Rich-Harris/estree-walker) to walk the tree, it’s what we use too, it’s light and already a peer dep so it should be much lighter than estraverse, and it can also replace
* You should be able to drop the clones, and change the tree directly, that should be a bunch faster
* I wonder what the best API is of this, for users? Should they get an array of components? An object of comments as keys to components as values? Could it make sense to change the default export (now an empty component) to that value?

**Reply by benjaminpreiss** [reply]
Alright, sure! Let me create an issue in our repo for this :)
